### ISSUE TITLE
Non-Github releases

### ISSUE THREAD

**Issue by drewcrawford** [Carthage/Carthage#722]
So

>  Carthage has been created as a decentralized dependency manager. There is no central list of projects, which reduces maintenance work and avoids any central point of failure.

However the "Releases" feature (#6) a.k.a. "pre-built frameworks", requires GitHub, a single point of failure.

I have been using a self-hosted GitLab lately, and I've been missing this feature.  Sadly, GitLab does not have something like Releases that carthage could be pointed to talk to, although they are [accepting PRs](http://feedback.gitlab.com/forums/176466-general/suggestions/4156755-releases-with-binary-files-like-on-github)...

I'm wondering however if it would be simpler and more compatible to invent a `.carthage-releases` file that sits in the repository root, which would just point to a binary hosted elsewhere:

```
$ cat .carthage-releases
0.6.8 => https://my-s3-bucket.s3.amazonaws.com/myframework-0.6.8.zip
0.6.9 => https://my-s3-bucket.s3.amazonaws.com/myframework-0.6.9.zip
etc.
```

...that way you can handle release hosting in all kinds of ways, and carthage would be compatible with all of them.

**Comment by michaelosity** [comment]
I just ran into this exact problem. +1

**Comment by flovilmart** [comment]
The releases could sit into the Cartfile itself, 

`release "https://my-s3-bucket.s3.amazonaws.com/myframework-0.6.8.zip"`

**Comment by mdiep** [comment]
Without a way to map an arbitrary release to a binary, I don't think there's much point. We shouldn't encourage people to tie themselves to a specific version just so they can use a binary.

**Comment by flovilmart** [comment]
I get your point, but in a certain way, that's quite impossible to get the sem-ver to work properly. Scanning the remote host for all URL's that match a version is quite problematic in an implementation point of view.
Unless we require a manifest on the server that lists versions, and full URL's.

In this case, that would be a simple, download, copy mechanism and we already have 99% of it.

Unless we force people to have a specific architecture on their host, I don't really see how it will be happening.

**Comment by drewcrawford** [comment]
I agree that pinning versions in the Cartfile itself is bad.  I do think the idea of using a manifest on the remote server has legs, e.g. for binary-only packages.

I also think we should explore putting the manifest in the git repository itself.  Checking a file into source control may be easier from a UX POV than maintaining some kind of folder structure on a host.

**Comment by flovilmart** [comment]
The manifest Would be The most lightweight solutions of all. If frameworks are hosted are binaries, maybe the developer doesn't want to open the repository itself in the first place.

**Comment by drewcrawford** [comment]
You're right.  Upon reflection I now agree with @flovilmart.

**Comment by neilpa** [comment]
This is treading into territory that's out of scope for Carthage. Closed source binaries should just be downloaded and linked by hand. You already have to do the latter for anyways for any Carthage built framework

**Comment by flovilmart** [comment]
I don't believe it's out of scope, as Github releases are supported.
Carthage is perfect for private repositories compared to cocoapods and I definitely see a usage that I could have today.

**Comment by cybertk** [comment]
_1

**Comment by flovilmart** [comment]
@mdiep on you comment about fixed versions, this is actually the proper way to managed dependencies. When using NPM, I make sure to use a properly specified version in order to make sure that upon deployment on servers I don't get an updated version that may break or behave differently.

Also, by specifying the full URL of the built framework, that allows more flexibility and less headache on carthage side as I mentioned earlier, we have 99.9% of that ready thanks to the logic of github releases.

We can also enforce that the frameworks should be build with the command `carthage build --no-skip-current` in order to avoid additional coding headaches.

**Comment by mdiep** [comment]
Okay, I think this is out of scope for Carthage. Attaching binaries to GitHub Releases makes sense because (a) we're already using that release information and (b) it's entirely automated.

But! The good news is that it should be possible to build this on top of Carthage. Carthage has been designed to expose CarthageKit as a usable framework. The design of CarthageKit itself still needs some work, but it should be possible to build something like this.

Otherwise you could take a look at https://github.com/guidomb/carthage_cache, which I think would satisfy your needs.

**Comment by drewcrawford** [comment]
Carthage is either a de-centralized package manager or it's not.

Closing this issue suggests that it's not.

**Comment by flovilmart** [comment]
Yeah it's a package manager that plays well with Github, first and foremost. I wonder why... Actually I don't...

**Comment by mdiep** [comment]
@flovilmart Passive aggressive comments aren't very effective. You're welcome to try to convince me that this is a good idea, but comments like that don't promote discussion.

> Carthage is either a de-centralized package manager or it's not.
> 
> Closing this issue suggests that it's not.

I don't think support for non-GitHub binary releases affects whether Carthage is decentralized. There's still no centralized system for specifying what is or isn't available via Carthage. That's what we mean by decentralized.

**Comment by flovilmart** [comment]
@mdiep ok , sorry for that comment, but you have to admit that supporting Github releases feature induce a lot of boilerplate and very specific code to `CarthageKit` for a particular feature that is not generic to `git`. This Github Releases feature is nice, we use it, but not all developers in our team have access to all repositories (for IP reasons). So we have to manually manage that particular framework, even if it's build with `carthage build --no-skip-current` and that it's own dependencies are carthage managed.

The point of *locking* to a certain version is actually best practices in development. All others package manager allows one to lock to a particular version.

When you compare HTTP hosted versions and Github releases, they actually share all the same code for retrieving the framework zip, as the Github Releases mechanism is actually asking the Github API for a URL and then downloading the specified files. I don't see the difference with something like `http https://my.secure.server.com/releases/Framework.zip`, maybe I am mistaken, but all the code is already written somewhere in CarthageKit to download a build framework from an http server right?

That's in that sense that I really don't understand why you dismiss it as 'out of scope' when a more complex feature like Github Releases has been implemented for a while.

All of that is great for open source projects but really painful for proprietary code, having to deal with 4 managements systems, Pods, Carthage, submodules and plain old copy pastes.

**Comment by drewcrawford** [comment]
I hope my comment wasn't interpreted as passive-agressive.

> I don't think support for non-GitHub binary releases affects whether Carthage is decentralized. There's still no centralized system for specifying what is or isn't available via Carthage.

What I thought a decentralized package management system means is that substantially all the features of the package manager (including e.g. binary releases) do not rely on a single point of failure (e.g. GitHub).

Under this decision, what a decentralized package manager means is "Technically, you can use some of Carthage without GitHub, but in practice we may make you a second-class citizen".  That has implications beyond just one issue.

Software projects have many reasons for not hosting here.  They may have infrastructure set up elsewhere, they may dislike GitHub as a company, they may go full Stallman and be trying to get rid of proprietary dependencies generally.

IMO, the "Just use GitHub" resolution implied here is a basic failure of the core mission.  I understand why it is technically easy to create this dependency for this feature but it does not align well with the decentralized package manager I want to be using.

**Comment by mdiep** [comment]
> I don't see the difference with something like `https://my.secure.server.com/releases/Framework.zip`, maybe I am mistaken, but all the code is already written somewhere in CarthageKit to download a build framework from an http server right?

The difference is in knowing what the URL is. In the case of GitHub Releases, you don't need to specify a URL, since we can retrieve it via the versioning information.

> What I thought a decentralized package management system means is that substantially all the features of the package manager (including e.g. binary releases) do not rely on a single point of failure (e.g. GitHub).

But Carthage already depends on GitHub for any project that's hosted on GitHub. If GitHub is down and you want to checkout https://github.com/ReactiveCocoa/ReactiveCocoa with Carthage, you'll be out of luck—with or without binary releases.

> IMO, the "Just use GitHub" resolution implied here is a basic failure of the core mission. I understand why it is technically easy to create this dependency for this feature but it does not align well with the decentralized package manager I want to be using.

I'm not completely opposed to supporting binary releases through some other means. But I don't think manifests are very workable.

I'll open this back up, but I think the way forward would be to specify a template for translating version numbers into binary release URLs. That would probably require `Cartfile`s to support more information, like the OGDL from #262.

**Comment by drewcrawford** [comment]
> But Carthage already depends on GitHub for any project that's hosted on GitHub. If GitHub is down and you want to checkout https://github.com/ReactiveCocoa/ReactiveCocoa with Carthage, you'll be out of luck—with or without binary releases.

The problem I have is not about *reliability*.  My concern are software projects who as an active choice don't use GitHub at all.  The problem isn't *GitHub goes down*.  The problem is that software projects may choose not to host there.  Carthage should not be a single-vendor show.

On a related note, GitLab [now has Release support](https://about.gitlab.com/2015/11/22/gitlab-8-2-released/), so adding that would bring us up to a two-vendor show.  I'm not sure that's the right thing to do, but it would solve my problem.

**Comment by jeeftor** [comment]
We are having issues because we cannot use github.  We have internal bitbucket and gitlab servers.  I would love to see this move forward

**Comment by flovilmart** [comment]
@mdiep We can do one or 2 things now because gitlab has releases support.

We can have the http distributed frameworks, packaged as MyFramework.zip, based on the current github release implementation. That should not take long. The drawback would be that the versioning would have to be handled by the URL itself, which is not that of a big deal as it's not the main use case.

Then another longer task would be to support Gitlab releases and implement their API.

**Comment by poborin** [comment]
Given a fact, that this feature is desired by many developers and there is no easy way to implement that into Carthage, I start thinking about plugins. If we have a clean plugins API we can write modules that will download artefacts from any repository.

Thus my question would be: is there any plugins API so far? If not, is there any plan for that?

**Comment by jeeftor** [comment]
There is def not a plugin api yet.

**Comment by poborin** [comment]
@jeeftor, that's actually not that bad. If we can specify in cartfile, let say that "this is a dependency, but we don't want to build that" it might help us to move forward towards the pure decentralised dependency manager.

```
git "git@my.cool:server/does-not-support-releases" --no-build
```

**Comment by grangej** [comment]
This would be amazing , since internally we must use stash, would love support for arbitrary binary support.

**Comment by tcurdt** [comment]
@mdiep Sorry, I don't get it. 

> The difference is in knowing what the URL is. In the case of GitHub Releases, you don't need to specify a URL, since we can retrieve it via the versioning information.

What is so bad about specifying meta information of a binary release in the `Cartfile`?

**Comment by mdiep** [comment]
There are at least a couple projects that will let you store binary releases on top of Carthage:

* [carthage_cache](https://github.com/guidomb/carthage_cache)
* [RomeBuild](https://github.com/146BC/RomeBuild)

Carthage uses Git tags for versioning information. GitHub Releases work for storing binary releases because they're built on top of Git tags.

We don't want to add a separate means of versioning. Doing so would add complexity and force framework authors to maintain duplicate information. That's antithetical to the Carthage vision of simplicity.

I see the projects above as the best solution for non-GitHub releases.

**Comment by tcurdt** [comment]
That reasoning I can maybe follow for source releases - not for binary ones though. When I tell Carthage "look - version 1.0.3 is at this URL" I don't see how that puts any burden on the framework author that did the release.

I guess that definition of _ethical_ and _simplicity_ is a bit of a disappointment (to me).
So it's either manual vendoring or back to CocoaPods (yay!).

**Comment by mdiep** [comment]
Want to open a separate issue to discuss binary-only releases?

**Comment by tcurdt** [comment]
@mdiep sure https://github.com/Carthage/Carthage/issues/1307

**Comment by sstadelman** [comment]
At SAP, we would really like to see this feature available, to let Carthage users update framework dependencies with Cartfile-based rules, as we publish updates.  We do have a [github presence](https://github.com/SAP), but our release policies dictate that only Open Source-licensed content can be distributed via the site.  Instead, we would prefer to make the frameworks available from our regular tooling site:  [https://tools.hana.ondemand.com/#mobile](https://tools.hana.ondemand.com/#mobile).  

Parsing the thread above, it looks like the main blocking issue is the lack of semantic versioning on unknown remote sources.  You point out that Github Releases *does* support this, which logically contributed to implementing the feature originally.  That makes sense.  But, maybe there are some acceptable limitations to versioning which could be unique to non-github sources?

A. Should remote sources *also* implement semantic versioning, according to the [Github API Releases](https://developer.github.com/v3/repos/releases/ ) spec?  If this were adopted, we could propose requiring the following services?
```
GET /:repo/releases
GET /:repo/releases/:id
GET /:repo/releases/latest
GET /:repo/releases/tags/:tag
```
B.  Scratch semantic versioning for remote sources.  The framework provider may version the `*.framework.zip` according to their own preference, but Carthage does no enforcement.  Ironically, the model for this is jQuery on the Google CDN, where [https://developers.google.com/speed/libraries/]("automatic version aliases") were deprecated, and developers should now supply full-qualified version URLs.
```
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.2/jquery.min.js"></script>
```
For our customers, we think that #2 would be quite acceptable.  (#1 might also be possible--and very nice).  #2 would be a short, pragmatic approach, and solve the basic problem of allowing *all* dependencies to be aggregated when using `carthage update`.  

Do either option make logical sense in the Carthage strategy?

**Comment by flovilmart** [comment]
@sstadelman for the simplicity of maintenance, I'd move for option B.

adding an:
```
http "https://url.to/myFramework.2.2.zip"
```

To the Cartfile seem the most sane and easy thing to do.

On another note, semantic versioning can be very unsafe for many reasons. I believe that the main use case is for private libraries, and people who know what they're doing.

The simplicity of deployment goes also for option B. Carthage historically is different from Cocoapods by the lack of centralized versioning information. Forcing developers to provide those endpoints can be really cumberstone and is not aligned with Carthage philosophy IMHO.

**Comment by mdiep** [comment]
@sstadelman It sounds like you're interested in distributing a binary-only framework—one for which the source code isn't available. Is that correct? If so, #1308 would be a better place to have this discussion. This issue is meant to discuss non-GitHub binary releases of frameworks where the source code is available.

**Comment by flovilmart** [comment]
@mdiep I don't believe it was ever intended to be non built frameworks:

``` 
However the "Releases" feature (#6) a.k.a. "pre-built frameworks", requires GitHub, a single point of failure.
```

Whenever we discussed it, I thought we were talking explicitly about built releases (through `carthage build --no-skip-current`)

**Comment by mdiep** [comment]
The existing Releases feature is an optimization. If a built/pre-built/binary release fails to download for whatever reason, then the source code will still be downloaded and Carthage will compile it. Those releases are completely optional for both the framework provider and the framework consumer.

Binary-only releases (where the source code isn't available to build) are a separate thing and should be discussed in #1308.

**Comment by simonbs** [comment]
At [Shape](http://www.shape.dk) we have discussed this issue. With large projects, we're finding that it takes a long time to update the dependencies. There are several reasons for this.

- Frameworks that do not support GitHub release, either because they are not hosted on GitHub or a release with a binary haven't been create, have to be compiled.
- We find that from time to time, Carthage freeze when fetching or cloning a repository, causing a need to restart the process.
- Debugging errors during fetching, cloning or compiling can sometimes be a long and cumbersome process.

Distributing binaries rather than source code will not solve all our issues but it will significantly speed up the process of updating dependencies for large projects. As most of our projects are hosted on Bitbucket, we are unable to take advantage of Carthages support for binary releases.

We have seen the approach proposed by @flovilmart in [a comment](https://github.com/Carthage/Carthage/issues/1308#issuecomment-227940441) for #1308 and the implementation proposed in #1363. While the approach is interesting and introducing a `http` keyword may make sense in order to support vendors that distribute a closed source framework, we do not find it sufficient for our requirements as the approach does not support semantic versioning.

We have discussed the issue intensively and as a supplement to GitHub Releases, we suggest adding the option to distribute binaries in a Git repository. We imagine the feature to work as follows.

1. As a framework author who desires to support binary releases, you archive your framework using `carthage archive`.
2. The resulting file, e.g. `MyFramework.framework.zip` is placed in the root of your repository and added to version control.
3. A tag describing the version of the framework is created, e.g. 1.0.2, and this is pushed to the remote.
3. When using the framework as a dependency and running `carthage update` or `carthage bootstrap`, Carthage will checkout the appropriate tag and look for the archived framework in the root of the repository. If the file is found, Carthage will treat this like it currently does with files retrieved through GitHub Releases.
4. If the file is not found, Carthage will continue compiling the source code.

We imagine this to be supported by the `git` keyword in Carthage, e.g. 

    git "git@bitbucket.org:shapedk/myframework.git" ~> 1.0.2

Naturally, the `—no-use-binaries` flag should still be respected.

While keeping a binary under version control might not seem ideal, we think this is a acceptable compromise as the approach has the benefit of working in arbitrary Git repositories, i.e. regardless of Git repository hosting service (GitHub, Bitbucket, GitLab, ...).

How does the approach sound? If it sounds reasonable, we will allot some time to attempt implementing it.

**Comment by jeeftor** [comment]
@simonbs 

We've managed to work around the carthage issue by using a CI server and building dependencies separately from the rest of the app (https://medium.com/@Jeef/speeding-up-carthage-builds-with-bamboo-90391917426b#.ta5afv4x4) Its still slow and very frustrating - but it helps somewhat.  

The approach, however doesn't fix the problem of having an internal project which uses an internal framework which uses other frameworks - that still takes some time.

**Comment by seivan** [comment]
What was wrong with using tags, but checking for a $FRAMEWORK_NAME.framework.zip  in the root directory, the same way carthage currently looks for schemes? 
Basically what  @simonbs suggested but it shouldn't have to require a whole git address.
```github "org/lib" ~> 1.0.2``` should still work

**Comment by simonbs** [comment]
@seivan `github "org/lib" ~> 1.0.2` would work with the current GitHub Releases. I'm using a Git address to support arbitrary Git repositories, primarily those not on GitHub.

**Comment by seivan** [comment]
Yup, I know, it’s just that it’s seems unnecessary to enforce using 'Github Release’ when a tag should suffice. This would make it more agnostic. I mean since you’re already suggestion look at root folder for $FRAMEWORK_NAME.framework.zip 
Might as well make the whole requirement of Github Release deprecated and just stick with regular git tags 'as uploading files' to Github makes it tricky to automate compared to regular git tags.

But yeah I agree with you, we’re on the same page of needing tags to work with binary releases.
